controllers/observability/perses.go

```diff
@@ -0,0 +1,215 @@
+		if err != nil {
+			return err
+		}
+		if entry.IsDir() {
+			return nil
+		}
+		if ext := filepath.Ext(entry.Name()); ext != ".yml" && ext != ".yaml" {
+			return nil
+		}
+		content, err := fs.ReadFile(d, path)
+		if err != nil {
+			return fmt.Errorf("failed to read %s: %w", path, err)
+		}
+		jsonBytes, err := yaml.YAMLToJSON(content)
+		if err != nil {
+			return fmt.Errorf("failed to convert yaml to json for %s: %w", path, err)
+		}
+		var objMap map[string]any
+		if err := json.Unmarshal(jsonBytes, &objMap); err != nil {
+			return fmt.Errorf("failed to unmarshal json for %s: %w", path, err)
+		}
+		out = append(out, objMap)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+func deepCopyMap(src map[string]any) map[string]any {
+	b, _ := json.Marshal(src)
+	var dst map[string]any
+	_ = json.Unmarshal(b, &dst)
+	return dst
+}
+
+// ReconcilePersesTokenSecret ensures a Secret with a bearer token usable by the HTTPProxy datasource
+// exists in the operator namespace. The token is requested for the "perses" ServiceAccount in the
+// namespace where a Perses instance is deployed.
+func (r *Reconciler) ReconcilePersesTokenSecret(ctx context.Context) error {
+	if !r.persesCRDsReady(ctx) {
+		return nil
+	}
+
+	// Discover Perses namespace by listing Perses CRs cluster-wide and picking the first.
+	var persesNs string
+	var persesList persesv1alpha1.PersesList
+	if err := r.List(ctx, &persesList); err == nil {
+		for _, p := range persesList.Items {
+			if p.Namespace != "" {
+				persesNs = p.Namespace
+				break
+			}
+		}
+	}
+	if persesNs == "" {
+		// Nothing to do if Perses is not installed
+		return nil
+	}
+
+	// Request a short-lived token for the "perses" SA in the Perses namespace
+	token, err := r.requestServiceAccountToken(ctx, persesNs, "perses")
+	if err != nil || token == "" {
+		// Do not fail reconcile; just skip creating the secret
+		return nil
+	}
+
+	// Create/Update secret in operator namespace
+	secret := &corev1.Secret{}
+	secret.Namespace = r.namespace
+	secret.Name = persesDatasourceSecretName
+	getErr := r.Get(ctx, types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}, secret)
+	data := map[string][]byte{"token": []byte(token)}
+	if getErr != nil {
+		if !apierrors.IsNotFound(getErr) {
+			// Unexpected error fetching the secret
+			return getErr
+		}
+		secret.Type = corev1.SecretTypeOpaque
+		secret.Data = data
+		return r.Create(ctx, secret)
+	}
+	// Update existing secret
+	if secret.Data == nil {
+		secret.Data = map[string][]byte{}
+	}
+	secret.Data["token"] = []byte(token)
+	return r.Update(ctx, secret)
+}
+
+// requestServiceAccountToken requests a token for the provided service account using TokenRequest API.
+func (r *Reconciler) requestServiceAccountToken(ctx context.Context, namespace, saName string) (string, error) {
+	clientset, err := kubernetes.NewForConfig(r.config)
```

Comment:
This should be done only once. Please add `clientset` as a new field to the Reconciler, and move this logic into the `NewReconciler` function.